File: docs/modules/examples/examples/clients/chat-room/go/consumer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
	"github.com/twmb/franz-go/pkg/kgo"
)

// tag::consumer[]
type Consumer struct {
	client *kgo.Client
	topic  string
}

// NewConsumer joins a new consumer group each run, starting from the
// beginning of the topic, so every client sees the whole chat history.
func NewConsumer(topic string) *Consumer {
	opts := append(clientOpts(),
		kgo.ConsumerGroup(uuid.New().String()),
		kgo.ConsumeTopics(topic),
		kgo.ConsumeResetOffset(kgo.NewOffset().AtStart()),
	)
	client, err := kgo.NewClient(opts...)
	if err != nil {
		panic(err)
	}
	return &Consumer{client: client, topic: topic}
}

// PrintMessages polls for records and prints each decoded message.
func (c *Consumer) PrintMessages() {
	for {
		fetches := c.client.PollFetches(context.Background())
		iter := fetches.RecordIter()
		for !iter.Done() {
			record := iter.Next()
			var msg Message
			if err := json.Unmarshal(record.Value, &msg); err != nil {
				fmt.Printf("Error decoding message: %v\n", err)
				continue
			}
			fmt.Printf("%s: %s\n", msg.User, msg.Message)
		}
	}
}
// end::consumer[]

func (c *Consumer) Close() {
	c.client.Close()
}

File: docs/modules/examples/examples/clients/chat-room/nodejs/src/consumer.ts
import { randomUUID } from "node:crypto";
import { redpanda, topic } from "./config";

// tag::consumer[]
// Each run joins a new consumer group and reads from the beginning of the
// topic, so every client sees the whole chat history.
const consumer = redpanda.consumer({
  kafkaJS: { groupId: randomUUID(), fromBeginning: true },
});

export async function connect() {
  await consumer.connect();
  await consumer.subscribe({ topics: [topic] });
  await consumer.run({
    eachMessage: async ({ message }) => {
      const { user, message: text } = JSON.parse(message.value!.toString());
      console.log(`${user}: ${text}`);
    },
  });
}
// end::consumer[]

export async function disconnect() {
  await consumer.disconnect();
}
